\NeedsTeXFormat{LaTeX2e}[1996/06/01]

\documentclass[11pt]{article}
\usepackage[rightcaption,raggedright]{sidecap}% for side captions
\usepackage{framed}         % for floatingboxes
\usepackage{soul}           % for letterspacing in theorem-style headings

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Special packages used AJD
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\usepackage{amsmath}

% for the Harvard author-date referencing system
%\usepackage[agsm]{harvard}

% if you are using either vancouver.bst or IEEEtran.bst and wish to remove
% square braces in the reference list, uncomment the line below
% \removesquarebraces

\usepackage{rotating}
\usepackage{floatpag}
\rotfloatpagestyle{empty}

% \usepackage{amsmath}% if you are using this package,
                      % it must be loaded before amsthm.sty
\usepackage{amsthm}
% \usepackage{txfonts}% times font (used to produce EngCguide.pdf)
                      % this package must be loaded after amsthm.sty
\usepackage{graphicx}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Special packages used AJD
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\usepackage{listings}

\lstset{
stringstyle=\ttfamily, % typewriter type for strings
showstringspaces=false % no special string spaces
}
\lstset{numbers=left, numberstyle=\tiny, stepnumber=1, numbersep=10pt, frame=trbl}
\lstset{escapeinside={(*@}{@*)}}

\usepackage{tikz,pgf}
  
\usetikzlibrary{positioning}


\def\sindex{}
\def\aindex{}

  
% BEAST book specific commands
\newcommand{\BEASTVersion}{2.1.x}
\newcommand{\TracerVersion}{1.6}
\newcommand{\FigTreeVersion}{1.4.2}
%\newcommand{\TODO}[1]{{\color{red} #1}}
\newcommand{\TODO}[1]{}
\newcommand{\chapter}[1]{}


% see chapter 3 for details
\theoremstyle{plain}% default
\newtheorem{theorem}{Theorem}
\newtheorem{lemma}[theorem]{Lemma}
\newtheorem*{corollary}{Corollary}

\theoremstyle{definition}
\newtheorem{definition}[theorem]{Definition}
\newtheorem{condition}[theorem]{Condition}

\theoremstyle{remark}
\newtheorem{remark}{Remark}
\newtheorem{notation}[remark]{Notation}
\newtheorem*{case}{Case}

\hyphenation{line-break line-breaks docu-ment triangle cambridge amsthdoc
cambridgemods baseline-skip author authors cambridgestyle en-vir-on-ment polar}

\setcounter{tocdepth}{2}% the toc normally lists sections;
% for the purposes of this document, this has been extended to subsections

%hyperref should come after index to make pagerefs work in index
\usepackage{hyperref}

\usepackage[all]{xy}


\title{Divergence Dating Tutorial with BEAST {\BEASTVersion}}
\author{Alexei Drummond, Andrew Rambaut, Remco Bouckaert, and Walter Xie}

\begin{document}
\maketitle

\newcommand{\chainLength}{{3,000,000}}
\newcommand{\logEvery}{{1,000}}
\newcommand{\screenEvery}{{10,000}}
\newcommand{\mccTree}{{\texttt{Primates.MCC.tree}}}

\section{Introduction}

This tutorial introduces the BEAST software for Bayesian evolutionary analysis through a simple tutorial. The tutorial involves co-estimation of a gene phylogeny and associated divergence times in the presence of calibration information from fossil evidence. 

You will need the following software at your disposal:

\begin{itemize}

\item {\bf BEAST} - this package contains the BEAST program, BEAUti, TreeAnnotator and other utility programs. This tutorial is written for BEAST v{\BEASTVersion}, which has support for multiple partitions. It is available for download from \texttt{http://www.beast2.org/}.
\item {\bf Tracer} - this program is used to explore the output of BEAST (and other Bayesian MCMC programs). It graphically and
quantitively summarizes the distributions of continuous parameters and provides diagnostic information. At the time of
writing, the current version is v{\TracerVersion}. It is available for download from\\*\texttt{http://tree.bio.ed.ac.uk/software/}.
\item {\bf FigTree} - this is an application for displaying and printing molecular phylogenies, in particular those obtained using
BEAST. At the time of writing, the current version is v{\FigTreeVersion}. It is available for download from \texttt{http://tree.bio.ed.ac.uk/software/}.
\end{itemize}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%
%%% TUTORIAL - RATES AND DATES
%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%\section{Rates and dates}

This tutorial will guide you through the analysis of an alignment of sequences sampled from twelve primate species (see Figure \ref{fig:primateAlignment}). The goal is to estimate the phylogeny as well as the rate of evolution on each lineage based on divergence times of their host species. 

\begin{figure}	

\includegraphics[width=\textwidth]{figures/AlignmentViewer.png}

\caption{Part of the alignment for primates.\label{fig:primateAlignment}}
\label{fig:BEAUti_ImportNexus}
\end{figure}

The first step will be to convert a NEXUS file with a DATA or CHARACTERS block into a BEAST XML input file. This is done using the program BEAUti (which stands for Bayesian Evolutionary Analysis Utility). 
This is a user-friendly program for setting the evolutionary model and options for the MCMC analysis. 
The second step is to actually run BEAST using the input file generated by BEAUTi,  which
contains the data, model and analysis settings. 
The final step is to explore the output of BEAST in order to diagnose problems and to summarize the results.

\section{BEAUti}

The program BEAUti is a user-friendly program for setting the
model parameters for BEAST. Run BEAUti by double clicking on its icon. Once running, \texttt{BEAUti} will look similar irrespective
of which computer system it is running on. For this tutorial, the Mac OS X version is used in the Figures but
the Linux and Windows versions will have the same layout and functionality.

\TODO{Provide instructions for executing BEAUti  in a Linux environment.}

\subsection{Loading the NEXUS file }

The example file called \texttt{primates-mtDNA.nex} is in the {\tt examples/nexus/} directory of the directory where BEAST was installed.
This file contains an alignment of sequences of 12 species of primates. 
%It looks likethis (the lines have been truncated):
%
%\begin{verbatim}
%#NEXUS
%begin data;
%dimensions ntax=12 nchar=898;
%format datatype=dna interleave=no gap=-;
%matrix
%Tarsius_syrichta  AAGTTTCATTGGAGCCACCACTCTTATAATTGCCCATGGCCTCACC
%Lemur_catta       AAGCTTCATAGGAGCAACCATTCTAATAATCGCACATGGCCTTACA
%Homo_sapiens	     AAGCTTCACCGGCGCAGTCATTCTCATAATCGCCCACGGGCTTACA
%Pan               AAGCTTCACCGGCGCAATTATCCTCATAATCGCCCACGGACTTACA
%Gorilla           AAGCTTCACCGGCGCAGTTGTTCTTATAATTGCCCACGGACTTACA
%Pongo             AAGCTTCACCGGCGCAACCACCCTCATGATTGCCCATGGACTCACA
%Hylobates         AAGCTTTACAGGTGCAACCGTCCTCATAATCGCCCACGGACTAACC
%Macaca_fuscata    AAGCTTTTCCGGCGCAACCATCCTTATGATCGCTCACGGACTCACC
%M_mulatta         AAGCTTTTCTGGCGCAACCATCCTCATGATTGCTCACGGACTCACC
%M_fascicularis    AAGCTTCTCCGGCGCAACCACCCTTATAATCGCCCACGGGCTCACC
%M_sylvanus        AAGCTTCTCCGGTGCAACTATCCTTATAGTTGCCCATGGACTCACC
%Saimiri_sciureus	 AAGCTTCACCGGCGCAATGATCCTAATAATCGCTCACGGGTTTACT
%;
%end;
%
%begin assumptions;
%	charset firsthalf = 1-449;
%	charset secondhalf = 450-898;
%end;
%end;\end{verbatim}
%
%\medskip{}

To load a NEXUS format alignment, simply select the \texttt{Import Alignment...} option from the File menu, or drag the file into the {\bf Partitions} panel. 

%\begin{figure}
%
%\includegraphics[width=\textwidth]{figures/ImportNexus}
%
%\caption{A screenshot of the import nexus file menu option in BEAUti. \TODO{Need to have the correct menu shortcuts for OS X!}}
%\label{fig:BEAUti_ImportNexus}
%\end{figure}

The following windows (Figure \ref{fig:addPartition} and \ref{fig:warning}) would appear if a related package is installed. If you are using ``pure'' BEAST 2, you can go to the next paragraph.    
If an \texttt{Add Partition} window pops up, you need to select \texttt{Add Alignment} and click \textbf{OK} to continue (Figure \ref{fig:addPartition}), and then read the following warning message to click \textbf{OK} (Figure \ref{fig:warning}).

\begin{figure}
\centering	
\includegraphics[width=7.5cm]{figures/AddPartition.png}
\caption{Add Partition window (Only appear if related packages are installed).}
\label{fig:addPartition}
\end{figure}

\begin{figure}	
\includegraphics[width=\textwidth]{figures/warning.png}
\caption{Warning message window (Only appear if related packages are installed).}
\label{fig:warning}
\end{figure}

Once loaded, five character partitions are displayed in the main panel (Figure \ref{fig:BEAUTI_DataPartitions}). {\bf You must remove the `coding' partition before continuing to the next step as it refers to the same nucleotides as partitions `1stpos', `2ndpos' and `3rdpos'.} To remove the `coding' partition select the row and click the `-' button at the bottom of the table. 
%You can view the alignment by double clicking the partition (Figure \ref{fig:BEAUTI_AlignmentViewer}).

\begin{figure}

\includegraphics[width=\textwidth]{figures/BEAUti_DataPartitions}
\caption{A screenshot of the data tab in BEAUti.}
\label{fig:BEAUTI_DataPartitions}
\end{figure}

%\begin{figure}[h]
%
%\includegraphics[width=\textwidth]{figures/AlignmentViewer}
%
%\caption{A screenshot of the alignment viewer in BEAUti.}
%\label{fig:BEAUTI_AlignmentViewer}
%\end{figure}

\subsection*{Link/Unlink partition models}

At this point we will need to link the clock model and tree. In the {\bf Partitions} panel, select all four partitions in the table (or none, by default all partitions are affected) and click the \texttt{Link Tree Models} button and then the \texttt{Link Clock Models} button (see Figure \ref{fig:BEAUti_DataPartitions_final}). Then click on the first drop-down menu in the Clock Model column and rename the shared clock model to `clock'. Likewise rename the shared tree to `tree'. This will make following options and generated log files more easy to read.

\begin{figure}

\includegraphics[width=\textwidth]{figures/BEAUti_DataPartitions_final}
\caption{A screenshot of the Partitions tab in BEAUti after linking and renaming the clock model and tree.}
\label{fig:BEAUti_DataPartitions_final}
\end{figure}

\subsection{Setting the substitution model}

The next step is to set up the substitution model. First we will temporarily link the site models in the Partitions panel so that we can change the model of all partitions simultaneously. Then, select the {\bf Site Models} tab at the top of the
main window. This will reveal the evolutionary model settings for
BEAST. The options available depend on whether the data are
nucleotides, or amino acids, binary data, or general data.
The settings that will appear after loading the primate nucleotide alignment will
be the default values for nucleotide data so we need to make some changes. 

Most of the models should be familiar to you. %(see Chapter {\ref{chapter:SubstitutionModels} for details}). 
First, set the \textbf{Gamma Category Count} to 4 and then check the `estimate' box for the \textbf{Shape} parameter. This will allow rate variation 
between sites in each partition to be modeled.  Then select  \textbf{HKY} from the \textbf{Subst Model} drop-down menu and select \textbf{Empirical} from the \textbf{Frequencies} drop-down menu. This will fix the frequencies to the proportions observed in the data (for each partition individually, once we unlink the site models again). This approach means that we can get a good fit to the data without explicitly estimating these parameters. We do it here simply to make the log files a bit shorter and more readable in later parts of the exercise. Finally check the `estimate' box for the \textbf{Substitution rate} parameter and select the \textbf{Fix mean mutation rate} check box. This will allow the individual partitions to have their relative rates estimated once we unlink the site models (Figure \ref{fig:BEAUti_Model}).

\begin{figure}
\includegraphics[width=\textwidth]{figures/BEAUti_Model}
\caption{A screenshot of the site model tab in BEAUti.}
\label{fig:BEAUti_Model}
\end{figure}

%Now, return to the {\bf Partitions} panel and unlink the site models so that each partition has its own named site model with independent substitution model parameters and relative rate.
Finally, hold `shift' key to select all site models on the left side, and click \textbf{OK} to clone the setting from \textbf{noncoding} into \textbf{1stpos}, \textbf{2ndpos} and \textbf{3rdpos} (Figure \ref{fig:cloneFrom}). 

\begin{figure}
\includegraphics[width=\textwidth]{figures/cloneFrom}
\caption{clone configuration from one site model to others.}
\label{fig:cloneFrom}
\end{figure}


\subsection{Setting the clock model}

The next step is to select the {\bf Clock Models} tab at the top of the
main window. This is where we select the molecular clock model. For this exercise we are going to leave the selection at the {\it default} value of a Strict molecular clock, because this data is very clock-like and does not need rate variation among branches to be included in the model.

%\begin{figure}
%\includegraphics[width=\textwidth]{figures/BEAUti_Clock}
%\caption{A screenshot of the clock model tab in BEAUti.}
%\label{fig:BEAUti_Clock}
%\end{figure}

\subsection{Priors }

The {\bf Priors} tab allows priors to be specified for each parameter in the
model. The model selections made in the site model and clock model tabs, result in the inclusion of various parameters
in the model, and these are shown in the priors tab (see Figure \ref{fig:BEAUti_Prior1}).

\begin{figure}
\includegraphics[width=\textwidth]{figures/BEAUti_Prior1}
\caption{A screenshot of the Priors tab in BEAUti. }
\label{fig:BEAUti_Prior1}
\end{figure}

Here we also specify that we wish to use the Calibrated Yule model \cite{Heled:2012fk}
as the tree prior. This is a simple model of speciation that
is generally more appropriate when considering sequences from different species. %(see Section \ref{section:YuleBirthDeathModels} for details).
Select this from the {\bf Tree prior} dropdown menu.

%\begin{figure}
%\includegraphics[width=\textwidth]{figures/BEAUti_Tree}
%\caption{A screenshot of the tree prior drop down menu in BEAUti.}
%\label{fig:BEAUti_Tree}
%\end{figure}

\subsubsection{Defining the calibration node}

To define an extra prior, press the small {\bf +} button below list of priors. You will see a
dialog that allows you to define a subset of the taxa in the phylogenetic tree. Once you have created a taxa set you will be able to add calibration information for its most recent common
ancestor (MRCA) later on. 

Name the taxa set by filling in the taxon set label entry. 
Call it \texttt{human-chimp} (it will contain the taxa for {\it Homo sapiens} and {\it Pan}).
In next list below you will see the available taxa. Select each of the two taxa in turn and press the $> >$ arrow button. (Figure \ref{fig:BEAUti_TaxonSets}).
Click OK and the newly defined taxa set will be added in to the prior list.
As this is a calibrated node to be used in conjunction with the Calibrated Yule prior, monophyly must be enforced, so select the checkbox marked \texttt{Monophyletic}. This will constrain the tree topology so that the human-chimp grouping is kept monophyletic during the course of the MCMC analysis.

\begin{figure}
\centering
\includegraphics[width=9cm]{figures/BEAUti_TaxonSets}
\caption{Taxon set editor in BEAUti.}
\label{fig:BEAUti_TaxonSets}
\end{figure}

To encode the calibration information we need to specify a distribution for the MRCA of human-chimp.
Select the \textbf{Log-normal} distribution from the drop down menu to the right of the newly added \texttt{human-chimp.prior}. 
Click on the black triangle and a graph of the probability density function will appear, along with parameters for the log normal distribution.
We are going to set $M=1.78$ and $S=0.085$ which will specify a distribution centred at about 6 million
years with a standard deviation of about 0.5 million years. This will give
a central 95\% probability range covering 5-7 Mya. This roughly corresponds to the current consensus
estimate of the date of the most recent common ancestor of humans and chimpanzees (Figure \ref{fig:BEAUti_Prior2}).

%We now need to specify a prior distribution on the calibrated node, based on our prior fossil knowledge. This is known
%as calibrating our tree. Select the \textbf{Log Normal} distribution from the drop down menu to the right of the newly added \texttt{human-%chimp.prior}. Click on the black triangle to the right and a graph of the probability density function will appear, along with parameters for the log-%normal distribution.
%We are going to specify a normal distribution centered at 6 million years with a standard deviation of 0.5 million years. This will give a central 95\% range of about 5-7 My. 
%This roughly corresponds to the current consensus estimate of the date of the most recent common ancestor of humans and chimpanzees (Figure %\ref{fig:BEAUti_TaxonSets}).

\begin{figure}
\includegraphics[width=\textwidth]{figures/BEAUti_Prior2}
\caption{A screenshot of the calibration prior options in the Priors panel in BEAUti.}
\label{fig:BEAUti_Prior2}
\end{figure}

Finally we will also specify some diffuse ``uninformative'' but proper priors on the overall molecular clock rate (\texttt{clockRate}) and the speciation rate (\texttt{birthRateY}) of the Yule tree prior. For each of them, select \textbf{Gamma} from the drop-down list and expand the view using the arrow button to the left. For both the clock rate and the Yule birth rate, set the Alpha (shape) parameter to 0.001 and the Beta (scale) parameter to 1000 (Figure \ref{fig:GammaPrior}).

\begin{figure}
\includegraphics[width=\textwidth]{figures/GammaPrior}
\caption{Gamma prior.}
\label{fig:GammaPrior}
\end{figure}

%Finally, put a log-normal distribution on the kappa parameters. \TODO{with what M and S parameters?}
%The clock model parameters will appear when the clock rate is estimated. 
%The priors table should now look like Figure \ref{fig:BEAUti_Prior2}.

%\begin{figure}
%\includegraphics[width=\textwidth]{figures/BEAUti_Prior2}
%\caption{A screenshot of the newly added calibration prior in the Priors panel in BEAUti. \TODO{This figure needs to be changed to remove the HomiCerco.prior.}}
%\label{fig:BEAUti_Prior2}
%\end{figure}

\subsection{Setting the MCMC options}

%Ignore the \textbf{Operators} tab as this just contains technical
%settings affecting the efficiency of the MCMC program (see Notes for details). 

The next tab, {\bf MCMC}, provides more general
settings to control the length of the MCMC run and the file names. 

Firstly we have the \textbf{Chain Length}. This is the number of
steps the MCMC will make in the chain before finishing. How long this
should be depends on the size of the data set, the complexity of the
model and the quality of answer required. The default value of 10,000,000
is entirely arbitrary and should be adjusted according to the size
of your data set. For this data set let's initially set the \texttt{chain
length} to \underline{\chainLength{}} as this will run reasonably quickly on most modern
computers (a few minutes).

%\begin{figure}
%\includegraphics[width=\textwidth]{figures/BEAUti_MCMC}
%\caption{A screenshot of the MCMC panel in BEAUti.}
%\label{fig:BEAUti_MCMC}
%\end{figure}

We will leave the \textbf{Store Every}  and \textbf{Pre Burnin} fields set to their default values. Below these are the details of the log files. Each one can be expanded by clicking the arrow to the right

The next options specify how often the parameter values in the Markov
chain should be displayed on the screen and recorded in the log file.
The screen output is simply for monitoring the programs progress so
can be set to any value (although if set too small, the sheer quantity
of information being displayed on the screen will actually slow the
program down). For the log file, the value should be set relative
to the total length of the chain. Sampling too often will result in
very large files with little extra benefit in terms of the accuracy
of the analysis. Sample too infrequently and the log file will not
record sufficient information about the distributions of the parameters. 
You probably want to aim to store no more than 10,000 samples so this should be
set to no less than chain length / 10,000.

For this exercise we will set the \texttt{screen log} to \underline{\screenEvery{}} and the \texttt{trace log} to \underline{\logEvery{}}. The final two
options give the file names of the log files for the sampled parameters and
the trees. These will be set to a default based on the name of the
imported NEXUS file. 

\begin{itemize}
\item If you are using the Windows operating system then we suggest you add the suffix \texttt{.txt} to both of these (so,
\texttt{Primates.log.txt} and \texttt{Primates.trees.txt}) so that Windows recognizes
these as text files. 
\end{itemize}

\subsection{Generating the BEAST XML file }

We are now ready to create the BEAST XML file. To do this,
select the \textbf{Save} option from the \textbf{File} menu. 
Check the default priors, and save the file with an appropriate name
(we usually end the filename with \texttt{.xml}, i.e., \texttt{Primates.xml}).
We are now ready to run the file through BEAST. 

\section{Running BEAST }

\begin{figure}
\includegraphics[width=\textwidth]{figures/BEAST}
\caption{A screenshot of BEAST.}
\label{fig:BEAST}
\end{figure}

Now run BEAST and when it asks for an input file, provide your newly
created XML file as input. BEAST will then run until it has finished
reporting information to the screen. The actual results files are
save to the disk in the same location as your input file. The output to the screen will
look something like this: 

{\scriptsize   
\begin{verbatim}
                  BEAST v2.1.3, 2002-2014
       Bayesian Evolutionary Analysis Sampling Trees
                 Designed and developed by
Remco Bouckaert, Alexei J. Drummond, Andrew Rambaut and Marc A. Suchard
                              
               Department of Computer Science
                   University of Auckland
                  remco@cs.auckland.ac.nz
                  alexei@cs.auckland.ac.nz
                              
             Institute of Evolutionary Biology
                  University of Edinburgh
                     a.rambaut@ed.ac.uk
                              
              David Geffen School of Medicine
           University of California, Los Angeles
                     msuchard@ucla.edu
                              
                Downloads, Help & Resources:
              	http://beast2.cs.auckland.ac.nz
                              
Source code distributed under the GNU Lesser General Public License:
              	http://code.google.com/p/beast2
                              
                     BEAST developers:
	Alex Alekseyenko, Trevor Bedford, Erik Bloomquist, Joseph Heled, 
	Sebastian Hoehna, Denise Kuehnert, Philippe Lemey, Wai Lok Sibon Li, 
	Gerton Lunter, Sidney Markowitz, Vladimir Minin, Michael Defoin Platel, 
          	Oliver Pybus, Chieh-Hsi Wu, Walter Xie
                              
                         Thanks to:
    	Roald Forsberg, Beth Shapiro and Korbinian Strimmer

File: Primates.xml seed: 777 threads: 1

  Using rescaling scheme : dynamic
  Using BEAGLE version: 2.1.2 resource 0: CPU
    with instance flags:  PRECISION_DOUBLE COMPUTATION_SYNCH EIGEN_REAL SCALING_MANUAL SCALERS_RAW VECTOR_NONE THREADING_NONE PROCESSOR_CPU FRAMEWORK_CPU
  Using BEAGLE version: 2.1.2 resource 0: CPU
    with instance flags:  PRECISION_DOUBLE COMPUTATION_SYNCH EIGEN_REAL SCALING_MANUAL SCALERS_RAW VECTOR_NONE THREADING_NONE PROCESSOR_CPU FRAMEWORK_CPU
  Using BEAGLE version: 2.1.2 resource 0: CPU
    with instance flags:  PRECISION_DOUBLE COMPUTATION_SYNCH EIGEN_REAL SCALING_MANUAL SCALERS_RAW VECTOR_NONE THREADING_NONE PROCESSOR_CPU FRAMEWORK_CPU
  Using BEAGLE version: 2.1.2 resource 0: CPU
    with instance flags:  PRECISION_DOUBLE COMPUTATION_SYNCH EIGEN_REAL SCALING_MANUAL SCALERS_RAW VECTOR_NONE THREADING_NONE PROCESSOR_CPU FRAMEWORK_CPU
======================================================
Please cite the following when publishing this model:

Remco Bouckaert, Joseph Heled, Denise Kuehnert, Tim Vaughan, Chieh-Hsi Wu, Dong Xie, Marc Suchard, Andrew Rambaut,Start likelihood: -7939.559067600949 
 Alexei J Drummond BEAST 2: A software platform for Bayesian evolutionary analysis. PLOS Computational Biology 10(4): e1003537, 2014

Heled J, Drummond AJ. Calibrated Tree Priors for Relaxed Phylogenetics and Divergence Time Estimation. Syst Biol (2012) 61 (1): 138-149.

Hasegawa, M., Kishino, H and Yano, T. 1985. Dating the human-ape splitting by a molecular clock of mitochondrial DNA. Journal of Molecular Evolution 22:160-174.


======================================================
Writing file /Users/dxie004/Documents/BEAST2/DivergenceDatingTutorial/primate-mtDNA.log
         Sample      posterior ESS(posterior)     likelihood          prior
Writing file /Users/dxie004/Documents/BEAST2/DivergenceDatingTutorial/tree.trees
              0     -7924.3599              N     -7688.4922      -235.8676 --
          10000     -5523.1281         2.0        -5451.4142       -71.7139 --
          20000     -5524.1632         3.0        -5445.1231       -79.0401 --
          30000     -5515.9203         4.0        -5441.8444       -74.0759 --
          40000     -5513.6938         5.0        -5440.5830       -73.1108 --
          50000     -5518.7113         6.0        -5447.3710       -71.3403 --
          60000     -5515.6182         7.0        -5440.9238       -74.6944 1m51s/Msamples
          70000     -5517.6669         8.0        -5444.8966       -72.7703 1m50s/Msamples
          ...
        2990000     -5519.5032       247.6        -5443.7829       -75.7202 1m55s/Msamples
        3000000     -5515.6591       249.3        -5444.7916       -70.8675 1m55s/Msamples
Operator                                                              Tuning	#accept	#reject	#total	acceptance rate
beast.evolution.operators.ScaleOperator_treeScaler.t:tree             0.694 	19418	87709	107127	0.181 
beast.evolution.operators.ScaleOperator_treeRootScaler.t:tree         0.644 	18839	88900	107739	0.175 
beast.evolution.operators.Uniform_UniformOperator.t:tree                    	240273	833290	1073563	0.224 
beast.evolution.operators.SubtreeSlide_SubtreeSlide.t:tree            9.098 	147088	390301	537389	0.274 
beast.evolution.operators.Exchange_narrow.t:tree                            	421	536829	537250	0.001 
beast.evolution.operators.Exchange_wide.t:tree                              	8	107544	107552	0 
beast.evolution.operators.WilsonBalding_WilsonBalding.t:tree                	8	106739	106747	0 
beast.evolution.operators.ScaleOperator_gammaShapeScaler.s:noncoding  0.389 	951	2502	3453	0.275 
beast.evolution.operators.ScaleOperator_KappaScaler.s:noncoding       0.387 	981	2491	3472	0.283 
beast.evolution.operators.DeltaExchangeOperator_FixMeanMutationRatesOperator0.381 	9591	61904	71495	0.134 
beast.evolution.operators.ScaleOperator_KappaScaler.s:1stpos          0.425 	825	2667	3492	0.236 
beast.evolution.operators.ScaleOperator_gammaShapeScaler.s:1stpos     0.386 	868	2794	3662	0.237 
beast.evolution.operators.ScaleOperator_KappaScaler.s:2ndpos          0.343 	924	2647	3571	0.259 
beast.evolution.operators.ScaleOperator_gammaShapeScaler.s:2ndpos     0.274 	867	2701	3568	0.243 
beast.evolution.operators.ScaleOperator_KappaScaler.s:3rdpos          0.551 	780	2719	3499	0.223 
beast.evolution.operators.ScaleOperator_gammaShapeScaler.s:3rdpos     0.333 	1133	2411	3544	0.32 
beast.evolution.operators.ScaleOperator_CalibratedYuleBirthRateScaler.t:tree0.260 	31511	75785	107296	0.294 
beast.evolution.operators.ScaleOperator_StrictClockRateScaler.c:clock 0.707 	25395	82605	108000	0.235 
beast.evolution.operators.UpDownOperator_strictClockUpDownOperator.c:clock0.592 	26181	81401	107582	0.243 
Total calculation time: 347.045 seconds
End likelihood: -5515.659153809371
\end{verbatim}}

\section{Analyzing the results}

Run the program called {\bf Tracer} to analyze the output of BEAST. When the main
window has opened, choose {\bf Import Trace File...} from the {\bf File} menu and select the file that
BEAST has created called \texttt{primate-mtDNA.log} (Figure \ref{fig:Tracer1}).

\begin{figure}
\includegraphics[width=\textwidth]{figures/Tracer1}
\caption{A screenshot of Tracer v{\TracerVersion}.}
\label{fig:Tracer1}
\end{figure}

Remember that MCMC is a stochastic algorithm so the actual numbers will not be exactly the same as those depicted in the figure.

On the left hand side is a list of the different quantities that BEAST has logged to file. 
There are traces for the posterior (this
is the natural logarithm of the product of the tree likelihood and the prior density), and the continuous parameters. Selecting a trace
on the left brings up analyses for this trace on the right hand side depending on tab that is selected. When first opened, the
`posterior' trace is selected and various statistics of this trace are shown under the Estimates tab.
In the top right of the window is a table of calculated statistics for the selected trace. 

Select the \texttt{clockRate} parameter in the lefthand list to look at
the average rate of evolution (averaged over the whole tree and all sites). Tracer will plot a (marginal posterior) histogram for the selected statistic and also give you
summary statistics such as the mean and median. The 95\% HPD stands for {\it highest posterior density interval} and represents the most compact interval on the selected parameter that contains 95\% of the posterior probability. It can be loosely thought of as a Bayesian analog to a confidence interval. The \texttt{TreeHeight} parameter gives the marginal posterior distribution of the age of the root of the entire tree.

Select the \texttt{TreeHeight} parameter and then Ctrl-click \texttt{mrcatime(human-chimp)}  (Command-click on Mac OS X). This will show a display of the
age of the root and the calibration MRCA we specified earlier in BEAUti. You can verify that the divergence that we used to calibrate the tree
(\texttt{mrcatime(human-chimp)}) has a posterior distribution that matches the prior distribution we specified (Figure \ref{fig:Tracer_divergences}).

\begin{figure}
\includegraphics[width=\textwidth]{figures/Tracer_divergences}
\caption{A screenshot of the 95\% HPD intervals of the root height and the user-specified (human-chimp) MRCA in Tracer.}
\label{fig:Tracer_divergences}
\end{figure}

\begin{figure}
\includegraphics[width=\textwidth]{figures/Tracer_marginalDensity}
\caption{A screenshot of the marginal posterior densities of the relative substitution rates of the four partitions (relative to the site-weighted mean rate). This plot shows that codon positions 1 and 2 have substantially different rates (0.455 versus 0.184) and both are far slower than codon position 3 with a relative rate of 2.942. The noncoding partition has a rate intermediate between codon positions 1 and 2 (0.345). Taken together this result suggests strong purifying selection in both the coding and noncoding regions of the alignment.}
\label{fig:Tracer_marginalDensity}
\end{figure}


\if0
\begin{figure}

\input{primatePriorPosteriorShape}
\caption{The marginal prior and posterior densities for the shape ($\alpha$) parameters. The prior is in gray. The posterior density estimate for each partition is also shown: noncoding (orange) and first (red), second (green) and third (blue) codon positions.}
\label{fig:primatePriorPosteriorShape}
\end{figure}

\begin{figure}
\input{primatePriorPosteriorKappa}
\caption{The marginal prior and posterior densities for the transition/tranversion bias ($\kappa$) parameters. The prior is in gray. The posterior density estimate for each partition is also shown: noncoding (orange) and first (red), second (green) and third (blue) codon positions.}
\label{fig:primatePriorPosteriorKappa}
\end{figure}
\fi

\subsection*{Questions}
\vspace{5 mm}

\textit{What is the estimated rate of molecular evolution for this gene tree (include the 95\% HPD interval)?}

\vspace{5 mm}
\framebox(420,30){}
\vspace{5 mm}

\textit{What sources of error does this estimate include?}

\vspace{5 mm}
\framebox(420,30){}
\vspace{5 mm}

%\item Does the rate of evolution differ substantially amongst different lineages in the tree?

\textit{How old is the root of the tree (give the mean and the 95\% HPD range)?}

\vspace{5 mm}
\framebox(420,30){}
\vspace{5 mm}
  
%The \texttt{coefficientOfVariation} statistic gives a summary of how much the rate of evolution varies from lineage to lineage (expressed as a proportion of the mean rate).
%   
%\bigskip{}
%
%Selecting the \texttt{treeModel.rootHeight} parameter gives the marginal posterior distribution of the age of the root of entire tree (i.e., the divergence between feline papillomavirus and canine oral %papillomavirus).
%
%\bigskip{}
%
%\textit{How old is the root of the tree (give the mean and the HPD range)?}
%
%\vspace{5 mm}
%\framebox(420,30){}
%\vspace{5 mm}


\section{Obtaining an estimate of the phylogenetic tree}

BEAST also produces a posterior sample of phylogenetic time-trees along with its sample of parameter estimates. 
These need to be summarized using the program {\bf TreeAnnotator}. This will take the set of trees and find the best
supported one. It will then annotate this representative summary tree with the mean ages of all the
nodes and the corresponding 95\% HPD ranges. It will also calculate the posterior clade probability for each
node. Run the TreeAnnotator program and set it up as depicted in Figure \ref{fig:TreeAnnotator1}.

\begin{figure}
\includegraphics[width=0.8\textwidth]{figures/TreeAnnotator1}
\caption{A screenshot of TreeAnnotator.}
\label{fig:TreeAnnotator1}
\end{figure}

The burnin is the number of trees to remove from the start of the sample. Unlike {\bf Tracer} which specifies the number of
steps as a burnin, in {\bf TreeAnnotator} you need to specify the actual number of trees. For this run, you specified a chain
length of \chainLength{} steps sampling every \logEvery{} steps. Thus the trees file will contain 3,000 trees and so to specify a 10\% burnin
in the top text field.

The {\bf Posterior probability limit} option specifies a limit such that if a node is found at less than this frequency in the sample
of trees (i.e., has a posterior probability less than this limit), it will not be annotated. The default of 0.5 means that only nodes
seen in the majority of trees will be annotated. Set this to zero to annotate all nodes.

For {\bf Target tree type} you can either choose a specific tree from a file or ask TreeAnnotator to find a tree in your sample.
The default option, {\bf Maximum clade credibility tree}, finds the tree with the highest product of the posterior probability of
all its nodes.

Choose {\bf Mean heights} for node heights. This sets the heights (ages) of each node in the tree to the mean height across the
entire sample of trees for that clade.

For the input file, select the trees file that BEAST created and select a file for the
output (here we called it \mccTree{}).

Now press Run and wait for the program to finish.

\begin{figure}
\includegraphics[width=0.8\textwidth]{figures/FigTree}
\includegraphics[width=0.7\textwidth]{figures/DensiTree}
\caption{A screenshot of FigTree and DensiTree.}
\label{fig:FigTree}
\end{figure}
\section{Visualizing the tree estimate}
Finally, we can visualize the tree in another program called {\bf FigTree}. Run this program, and open
the \mccTree{} file by using the Open command in the File menu. The tree should appear.
You can now try selecting some of the options in the control panel on the left. Try selecting
{\bf Node Bars} to get node age error bars. Also turn on {\bf Branch Labels} and select {\bf posterior} to get
it to display the posterior probability for each node. 
If you use a non strict clock model then under {\bf Appearance} you can also tell FigTree to colour the branches by the rate.
You should end up with something similar to Figure \ref{fig:FigTree}.


An alternative view of the tree can be made with DensiTree, which is part of Beast 2. The advantage
of DensiTree is that it is able to visualize both uncertainty in node heights and uncertainty in topology.
For this particular dataset, the dominant topology is present in more than 99\% of the samples. So, 
we conclude that this analysis results in a very high consensus on topology (Figure \ref{fig:FigTree}).


%\begin{figure}
%\begin{center}
%\includegraphics[width=0.7\textwidth]{figures/DensiTree}
%\end{center}
%\caption{A screenshot of DensiTree.}
%\label{fig:DensiTree}
%\end{figure}

\subsection*{Questions}
\vspace{5 mm}

\textit{Does the rate of evolution differ substantially amongst different lineages in the tree?}

\vspace{5 mm}
\framebox(420,60){}
\vspace{5 mm}

DensiTree has a clade bar (Menu Window/View clade toolbar) to show information on clades.

\textit{What is the support for the clade [Homo\_sapiens, Pan, Gorilla, Hylobates]?}

\vspace{5 mm}
\framebox(420,30){}
\vspace{5 mm}

You can browse through the topologies in DensiTree using the Browse menu.
The most popular topology has a support of over 99\%.

\textit{What is the support for the second most popular topology?}

\vspace{5 mm}
\framebox(420,30){}
\vspace{5 mm}

Under the help menu, DensiTree shows some information.

\textit{How many topologies are in the tree set?}

\vspace{5 mm}
\framebox(420,30){}
\vspace{5 mm}

\section{Comparing your results to the prior}

Using BEAUti, set up the same analysis but under the MCMC options, select the {\bf Sample from prior only} option. This will allow you to visualize the full prior distribution in the absence of your sequence data. Summarize the trees from the full prior
distribution and compare the summary to the posterior summary tree.




% \renewcommand{\refname}{Bibliography}% if you prefer this heading
\bibliographystyle{amsplain} 
\bibliography{DivergenceDatingTutorial}








%\newpage


\end{document}


